tok/hnsw/helper.go

```diff
@@ -544,13 +595,33 @@ func (ph *persistentHNSW[T]) addNeighbors(ctx context.Context, tc *TxnCache,
 			}
 		}
 	}
+	var inVec, outVec []T
 	for level := 0; level < ph.maxLevels; level++ {
 		allLayerEdges[level], nnEdgesErr = ph.removeDeadNodes(allLayerEdges[level], tc)
 		if nnEdgesErr != nil {
 			return nil, nnEdgesErr
 		}
 		// This adds at most efConstruction number of edges for each layer for this node
 		allLayerEdges[level] = append(allLayerEdges[level], allLayerNeighbors[level]...)
+		if len(allLayerEdges[level]) > ph.efConstruction {
+			err := ph.getVecFromUid(uuid, tc, &inVec)
+			if err != nil {
+				log.Printf("[ERROR] While getting vector %s", err)
+			} else {
+				h := &CustomHeap{
+					data: allLayerEdges[level],
+					compare: func(i, j uint64) bool {
+						return ph.distance_betw(ctx, tc, uuid, i, &inVec, &outVec) >
```

Comment:
we're still computing vectors distances at every insert addNeighbors call. instead, once the neighbors are calculated, the edges should now store the distance between the two nodes. so youll  need to modify what nodeAllEdges gives you.

Reply:
make that a heap instead that gets autosorted and you can just pop the worst one at addNeighbor time, and its O(1) cost.